File: java/src/main/java/dev/drim/relay/app/AttachmentsController.java
package dev.drim.relay.app;

import dev.drim.relay.domain.Entities;
import dev.drim.relay.domain.Role;
import dev.drim.relay.id.IdFactory;
import dev.drim.relay.seams.AttachmentAccess;
import dev.drim.relay.seams.AttachmentStore;
import dev.drim.relay.seams.ChannelRoleGate;
import dev.drim.relay.store.AttachmentRepository;
import dev.drim.relay.store.entity.AttachmentRow;
import dev.drim.relay.web.ApiException;
import dev.drim.relay.web.CurrentUser;
import java.io.IOException;
import java.time.Instant;
import org.springframework.core.io.ByteArrayResource;
import org.springframework.core.io.Resource;
import org.springframework.http.ContentDisposition;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

/**
 * Attachments: upload (member+, ≤ 1 MiB, non-empty) writes bytes to the object store under an
 * opaque key then records the metadata row; download authorizes via {@link AttachmentAccess} (G-S3
 * — membership, never key possession) and streams the bytes back.
 */
@RestController
public class AttachmentsController {
  private static final long MAX_ATTACHMENT_BYTES = 1L << 20; // 1 MiB

  private final CurrentUser currentUser;
  private final ChannelRoleGate roleGate;
  private final AttachmentAccess attachmentAccess;
  private final AttachmentStore store;
  private final AttachmentRepository attachments;
  private final IdFactory ids;

  public AttachmentsController(
      CurrentUser currentUser,
      ChannelRoleGate roleGate,
      AttachmentAccess attachmentAccess,
      AttachmentStore store,
      AttachmentRepository attachments,
      IdFactory ids) {
    this.currentUser = currentUser;
    this.roleGate = roleGate;
    this.attachmentAccess = attachmentAccess;
    this.store = store;
    this.attachments = attachments;
    this.ids = ids;
  }

  public record AttachmentDto(
      String id, String channelId, String filename, long sizeBytes, Instant createdAt) {}

  @PostMapping("/channels/{id}/attachments")
  public ResponseEntity<AttachmentDto> upload(
      @PathVariable String id, @RequestParam(name = "file", required = false) MultipartFile file) {
    roleGate.authorizeRole(id, currentUser.id(), Role.MEMBER);

    if (file == null || file.isEmpty()) {
      if (file == null) {
        throw ApiException.invalid("attachment:invalid", "A file field is required.");
      }
      throw ApiException.invalid("attachment:empty", "The attachment is empty.");
    }
    if (file.getSize() > MAX_ATTACHMENT_BYTES) {
      throw ApiException.tooLarge(
          "attachment:too_large", "The attachment exceeds the 1 MiB limit.");
    }
    byte[] content;
    try {
      content = file.getBytes();
    } catch (IOException e) {
      throw ApiException.invalid("attachment:invalid", "Could not read the upload.");
    }
    if (content.length > MAX_ATTACHMENT_BYTES) {
      throw ApiException.tooLarge(
          "attachment:too_large", "The attachment exceeds the 1 MiB limit.");
    }
    if (content.length == 0) {
      throw ApiException.invalid("attachment:empty", "The attachment is empty.");
    }

    String attachmentId = ids.create();
    String storageKey = id + "/" + attachmentId;
    store.put(storageKey, content);
    AttachmentRow row =
        new AttachmentRow(
            attachmentId,
            id,
            currentUser.id(),
            null,
            file.getOriginalFilename(),
            content.length,
            storageKey,
            Instant.now());
    attachments.save(row);

    return ResponseEntity.status(HttpStatus.CREATED)
        .body(
            new AttachmentDto(
                row.getId(),
                row.getChannelId(),
                row.getFilename(),
                row.getSizeBytes(),
                row.getCreatedAt()));
  }

  @GetMapping("/attachments/{id}")
  public ResponseEntity<Resource> download(@PathVariable String id) {
    Entities.Attachment attachment = attachmentAccess.authorize(id, currentUser.id());
    byte[] content = store.get(attachment.storageKey());
    return ResponseEntity.ok()
        .contentType(MediaType.APPLICATION_OCTET_STREAM)
        .header(
            HttpHeaders.CONTENT_DISPOSITION,
            ContentDisposition.attachment().filename(attachment.filename()).build().toString())
        .body(new ByteArrayResource(content));
  }
}

File: typescript/src/app/attachments.controller.ts
import {
  Controller,
  Get,
  HttpCode,
  Inject,
  Param,
  Post,
  Req,
  Res,
  UploadedFile,
  UseInterceptors,
} from '@nestjs/common';
import { FileInterceptor } from '@nestjs/platform-express';
import type { Response } from 'express';

import { invalid, tooLarge } from '../apierr/apierr.js';
import type { Attachment } from '../domain/domain.js';
import { Role } from '../domain/domain.js';
import { IdFactory } from '../idgen/idgen.js';
import {
  ATTACHMENT_ACCESS,
  ATTACHMENT_STORE,
  CHANNEL_ROLE_GATE,
  type AttachmentAccess,
  type AttachmentStore,
  type ChannelRoleGate,
} from '../seams/seams.js';
import { Store } from '../store/store.js';
import { currentUser, type RequestWithUser } from './identity.middleware.js';

const MAX_ATTACHMENT_BYTES = 1 << 20; // 1 MiB

interface UploadedMulterFile {
  originalname: string;
  buffer: Buffer;
  size: number;
}

interface AttachmentDto {
  id: string;
  channelId: string;
  filename: string;
  sizeBytes: number;
  createdAt: string;
}

@Controller()
export class AttachmentsController {
  constructor(
    private readonly store: Store,
    private readonly ids: IdFactory,
    @Inject(CHANNEL_ROLE_GATE) private readonly roleGate: ChannelRoleGate,
    @Inject(ATTACHMENT_STORE) private readonly objects: AttachmentStore,
    @Inject(ATTACHMENT_ACCESS) private readonly access: AttachmentAccess,
  ) {}

  @Post('channels/:id/attachments')
  @HttpCode(201)
  // Multer caps the field at 1 MiB + 1 so an over-limit upload is detected, not
  // truncated; the handler turns the cap and the empty case into the pinned codes.
  @UseInterceptors(FileInterceptor('file', { limits: { fileSize: MAX_ATTACHMENT_BYTES + 1 } }))
  async upload(
    @Req() req: RequestWithUser,
    @Param('id') channelId: string,
    @UploadedFile() file?: UploadedMulterFile,
  ): Promise<AttachmentDto> {
    const caller = currentUser(req);
    await this.roleGate.authorizeRole(channelId, caller.id, Role.Member);
    if (!file) {
      throw invalid('attachment:invalid', 'A file field is required.');
    }
    if (file.size > MAX_ATTACHMENT_BYTES) {
      throw tooLarge('attachment:too_large', 'The attachment exceeds the 1 MiB limit.');
    }
    if (file.size === 0) {
      throw invalid('attachment:empty', 'The attachment is empty.');
    }
    const id = this.ids.create();
    const storageKey = `${channelId}/${id}`;
    await this.objects.put(storageKey, file.buffer);
    const att: Attachment = {
      id,
      channelId,
      uploaderId: caller.id,
      messageId: null,
      filename: file.originalname,
      sizeBytes: file.size,
      storageKey,
      createdAt: new Date(),
    };
    await this.store.insertAttachment(att);
    return { id: att.id, channelId, filename: att.filename, sizeBytes: att.sizeBytes, createdAt: att.createdAt.toISOString() };
  }

  @Get('attachments/:id')
  async download(@Req() req: RequestWithUser, @Res() res: Response, @Param('id') id: string): Promise<void> {
    const caller = currentUser(req);
    const att = await this.access.authorize(id, caller.id);
    const content = await this.objects.get(att.storageKey);
    res.setHeader('Content-Type', 'application/octet-stream');
    res.setHeader('Content-Disposition', `attachment; filename="${att.filename}"`);
    res.status(200).send(content);
  }
}
